python/tskit/metadata.py

```diff
@@ -0,0 +1,115 @@
+# MIT License
+#
+# Copyright (c) 2018-2019 Tskit Developers
+# Copyright (c) 2017 University of Oxford
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Classes for metadata decoding, encoding and validation
+"""
+import abc
+import json
+from typing import Any
+
+import jsonschema
+
+import tskit.exceptions as exceptions
+
+
+class MetadataCodec(abc.ABC):
+    @abc.abstractmethod
+    def decode(self, encoded_bytes: bytes) -> Any:
+        pass
+
+    @abc.abstractmethod
+    def encode(self, obj: Any) -> bytes:
+        pass
+
+
+class JSONCodec(MetadataCodec):
```

Comment:
Tending to use e.g. ``JsonCodec`` elsewhere in the code, but I don't feel strongly about it.

Reply:
I was following the python way e.g. https://docs.python.org/3/library/json.html#json.JSONEncoder I'll change the others in the code?